[pkg/stanza/adapter] Missing storage extension causes panic during shutdown

### Component(s)

pkg/stanza

### What happened?

## Steps to Reproduce
```yaml
receivers:
  filelog:
    storage: foo
    include: foo.log

exporters:
  logging:

service:
  pipelines:
    logs:
      receivers: [filelog]
      exporters: [logging]
```

## Expected Result
```
2022-11-09T10:23:14.838-0500    info    service/service.go:129  Shutdown complete.
Error: cannot start pipelines: storage client: storage extension 'foo' not found
2022-11-09 10:23:14.838221 I | collector server run finished with error: cannot start pipelines: storage client: storage extension 'foo' not found
```

## Actual Result
```
2022-11-09T10:22:15.859-0500    info    adapter/receiver.go:145 Stopping stanza receiver        {"kind": "receiver", "name": "filelog", "pipeline": "logs"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x20 pc=0x105c78b9c]

goroutine 1 [running]:
github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/adapter.(*receiver).Shutdown(0x1400101a6c0, {0x109df1ee0, 0x1400019c000})
        github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza@v0.64.0/adapter/receiver.go:151 +0x9c
go.opentelemetry.io/collector/service/internal/pipelines.(*Pipelines).ShutdownAll(0x140000c9090, {0x109df1ee0, 0x1400019c000})
        go.opentelemetry.io/collector@v0.64.0/service/internal/pipelines/pipelines.go:121 +0x378
go.opentelemetry.io/collector/service.(*service).Shutdown(0x14000acce80, {0x109df1ee0, 0x1400019c000})
```

### Collector version

v0.63.1

### Environment information

_No response_

### OpenTelemetry Collector configuration

_No response_

### Log output

_No response_

### Additional context

_No response_

--- a/.chloggen/pkg-stanza-missing-storage-panic.yaml
+++ b/.chloggen/pkg-stanza-missing-storage-panic.yaml
@@ -1,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: pkg/stanza, filelog, journald, windowseventlog
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix issue where specifying a non-existent storage extension caused panic during shutdown.
+
+# One or more tracking issues related to the change
+issues: [16212]
+
+# (Optional) One or more lines of additional information to render under the primary note.
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext:

--- a/pkg/stanza/adapter/receiver.go
+++ b/pkg/stanza/adapter/receiver.go
@@ -148,6 +148,9 @@ func (r *receiver) Shutdown(ctx context.Context) error {
 	r.cancel()
 	r.wg.Wait()
 
-	clientErr := r.storageClient.Close(ctx)
-	return multierr.Combine(pipelineErr, clientErr)
+	if r.storageClient != nil {
+		clientErr := r.storageClient.Close(ctx)
+		return multierr.Combine(pipelineErr, clientErr)
+	}
+	return pipelineErr
 }

--- a/pkg/stanza/adapter/storage_test.go
+++ b/pkg/stanza/adapter/storage_test.go
@@ -99,6 +99,7 @@ func TestFailOnMissingStorageExtension(t *testing.T) {
 	err := r.Start(context.Background(), storagetest.NewStorageHost())
 	require.Error(t, err)
 	require.Equal(t, "storage client: storage extension 'test/missing' not found", err.Error())
+	require.NoError(t, r.Shutdown(context.Background()))
 }
 
 func TestFailOnNonStorageExtension(t *testing.T) {

